platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/DefaultProblemReport.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.problems.internal;
+
+import org.gradle.api.NonNullApi;
+import org.gradle.api.problems.Severity;
+
+import javax.annotation.Nullable;
+import java.io.Serializable;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+@NonNullApi
+public class DefaultProblemReport implements Serializable, ProblemReport {
+    private final ProblemDefinition problemDefinition;
+    private final ProblemContext problemContext;
+
+
+    protected DefaultProblemReport(
+        String label,
+        Severity severity,
+        List<ProblemLocation> locations,
+        @Nullable DocLink documentationUrl,
+        @Nullable String description,
+        @Nullable List<String> solutions,
+        @Nullable RuntimeException cause,
+        ProblemCategory problemCategory,
+        Map<String, Object> additionalData
```

Comment:
🤔 I find it weird that we introduced these two concepts, but here, we need to supply all the sub-components instead of getting an already-made definition and context.

Reply:
That probably is something I should cover in this PR. Will look into it.